cli/tools/test_runner.rs

```diff
@@ -63,42 +108,295 @@ pub fn prepare_test_modules_urls(
+          test_module.clone(),
+          permissions.clone(),
+          sender.clone(),
+        );
+
+        tokio_util::run_basic(future)
+      });
+
+      join_handle.join().unwrap()
+    })
+  });
+
+  let join_futures = stream::iter(join_handles)
+    .buffer_unordered(concurrent_jobs)
+    .collect::<Vec<Result<Result<(), AnyError>, tokio::task::JoinError>>>();
+
+  let handler = {
+    tokio::task::spawn_blocking(move || {
+      let time = std::time::Instant::now();
+      let mut failed = 0;
+      let mut filtered_out = 0;
+      let mut ignored = 0;
+      let mut passed = 0;
+      let measured = 0;
+
+      let mut used_only = false;
+      let mut has_error = false;
+      let mut failures: Vec<(String, String)> = Vec::new();
+
+      for message in receiver.iter() {
+        match message {
+          TestMessage::Plan {
+            pending,
+            filtered,
+            only,
+          } => {
+            println!("running {} tests", pending);
+
+            if only {
+              used_only = true;
+            }
+
+            filtered_out += filtered;
+          }
+
+          TestMessage::Result {
+            name,
+            duration,
+            result,
+          } => match result {
+            TestResult::Ok => {
+              println!(
+                "test {} ... {} {}",
+                name,
+                colors::green("ok"),
+                colors::gray(format!("({}ms)", duration))
+              );
+
+              passed += 1;
+            }
+            TestResult::Ignored => {
+              println!(
+                "test {} ... {} {}",
+                name,
+                colors::yellow("ignored"),
+                colors::gray(format!("({}ms)", duration))
+              );
+
+              ignored += 1;
+            }
+            TestResult::Failed(error) => {
+              println!(
+                "test {} ... {} {}",
+                name,
+                colors::red("FAILED"),
+                colors::gray(format!("({}ms)", duration))
+              );
+
+              failed += 1;
+              failures.push((name, error));
+              has_error = true;
+            }
+          },
+          _ => {}
+        }
+
+        if has_error && fail_fast {
+          break;
+        }
+      }
+
+      if !failures.is_empty() {
+        println!("\nfailures:\n");
+        for (name, error) in &failures {
+          println!("{}", name);
+          println!("{}", error);
+          println!();
+        }
+
+        println!("failures:\n");
+        for (name, _) in &failures {
+          println!("\t{}", name);
+        }
+      }
+
+      let status = if failures.is_empty() {
+        colors::green("ok").to_string()
+      } else {
+        colors::red("FAILED").to_string()
+      };
+
+      println!(
+        "\ntest result: {}. {} passed; {} failed; {} ignored; {} measured; {} filtered out {}\n",
+        status,
+        passed,
+        failed,
+        ignored,
```

Comment:
Just to confirm - this output is matching the currently existing one, correct?

Reply:
Yup, to the tee including the ugly bits (raising that issue in a different PR).

Reply:
An aside; this is ugly but have a branch where this whole block except the "control" bits (has_error, used_only) of the thread it are going into a `TestReporter`.